File: elixir/9-PalindromeNumber/main.exs
defmodule Solution do
    @spec reverse(x :: integer, acc :: integer) :: integer
    defp reverse(x, acc) do
        if x == 0 do
            acc
        else
            reverse(div(x, 10), (acc * 10) + rem(x, 10)) 
        end
    end

    @spec is_palindrome(x :: integer) :: boolean
    def is_palindrome(x) when x < 0 do
        false
    end

    @spec is_palindrome(x :: integer) :: boolean
    def is_palindrome(x) when x == 0 do
        true
    end
  
    @spec is_palindrome(x :: integer) :: boolean
    def is_palindrome(x) do
        reverse(x, 0) == x
    end
end

IO.puts(Solution.is_palindrome(121))
IO.puts(Solution.is_palindrome(-121))
IO.puts(Solution.is_palindrome(10))

File: cpp/9-PalindromeNumber/main.cpp
#include <iostream>

using namespace std;

class Solution 
{
public:
    bool isPalindrome(int x) 
    {
        if(x < 0) return false;

        auto original = x;
        uint64_t number = 0;

        while(x)
        {
            number *= 10;
            number += (x % 10);

            x /= 10;
        }

        return number == original;
    }
};

int main()
{
    Solution s;
    cout << s.isPalindrome(10) << endl;
    cout << s.isPalindrome(12321) << endl;

    return 0;
}